File: backend/src/main/java/com/schoolfinder/service/SchoolMapper.java
package com.schoolfinder.service;

import com.schoolfinder.api.Dtos.ImageDto;
import com.schoolfinder.api.Dtos.ProgramDto;
import com.schoolfinder.api.Dtos.SchoolDetail;
import com.schoolfinder.api.Dtos.SchoolSummary;
import com.schoolfinder.domain.School;
import java.util.List;

public final class SchoolMapper {

    private SchoolMapper() {
    }

    public static SchoolSummary toSummary(School s, Rating rating) {
        return new SchoolSummary(
            s.getId(),
            s.getName(),
            s.getCategory(),
            s.getCity(),
            s.getRegion(),
            s.getTuitionFee(),
            s.getCurrency(),
            s.getCoverImageUrl(),
            s.getLatitude(),
            s.getLongitude(),
            rating.average(),
            rating.count()
        );
    }

    public static SchoolDetail toDetail(School s, Rating rating, boolean isFavorite) {
        List<ProgramDto> programs = s.getPrograms().stream()
            .map(p -> new ProgramDto(p.getId(), p.getName(), p.getFaculty(), p.getLevel(), p.getDurationMonths(), p.getTuitionFee()))
            .toList();
        List<ImageDto> images = s.getImages().stream()
            .map(i -> new ImageDto(i.getId(), i.getUrl(), i.getCaption()))
            .toList();
        return new SchoolDetail(
            s.getId(),
            s.getName(),
            s.getCategory(),
            s.getDescription(),
            s.getHistory(),
            s.getCity(),
            s.getRegion(),
            s.getAddress(),
            s.getLatitude(),
            s.getLongitude(),
            s.getTuitionFee(),
            s.getCurrency(),
            s.getWebsite(),
            s.getPhone(),
            s.getEmail(),
            s.getCoverImageUrl(),
            rating.average(),
            rating.count(),
            isFavorite,
            programs,
            images
        );
    }
}

File: backend/src/main/kotlin/com/schoolfinder/service/SchoolMapper.kt
package com.schoolfinder.service

import com.schoolfinder.api.ImageDto
import com.schoolfinder.api.ProgramDto
import com.schoolfinder.api.SchoolDetail
import com.schoolfinder.api.SchoolSummary
import com.schoolfinder.domain.School

/** Rating numbers for a school, rounded to one decimal place. */
data class Rating(val average: Double, val count: Long)

object SchoolMapper {

    fun toSummary(s: School, rating: Rating) = SchoolSummary(
        id = s.id!!,
        name = s.name,
        category = s.category,
        city = s.city,
        region = s.region,
        tuitionFee = s.tuitionFee,
        currency = s.currency,
        coverImageUrl = s.coverImageUrl,
        latitude = s.latitude,
        longitude = s.longitude,
        averageRating = rating.average,
        ratingCount = rating.count,
    )

    fun toDetail(s: School, rating: Rating, isFavorite: Boolean) = SchoolDetail(
        id = s.id!!,
        name = s.name,
        category = s.category,
        description = s.description,
        history = s.history,
        city = s.city,
        region = s.region,
        address = s.address,
        latitude = s.latitude,
        longitude = s.longitude,
        tuitionFee = s.tuitionFee,
        currency = s.currency,
        website = s.website,
        phone = s.phone,
        email = s.email,
        coverImageUrl = s.coverImageUrl,
        averageRating = rating.average,
        ratingCount = rating.count,
        favorite = isFavorite,
        programs = s.programs.map { ProgramDto(it.id!!, it.name, it.faculty, it.level, it.durationMonths, it.tuitionFee) },
        images = s.images.map { ImageDto(it.id!!, it.url, it.caption) },
    )
}
